### PR TITLE
Feat/bba handler bval

### LINKED ISSUE #84: [bba] handle bval
**Subject**
@zeroFruit 

**Proposal Detail**
impl bval handler

### PR DESCRIPTION
**Pull request by zeroFruit**
## Related Issue

resolve: #84

## Description

impl bval handler

## Checklist

Thank you for your contribution.
Before submitting this PR, please make sure:

- [x] Applying goimports
- [x] Test case
- [x] End of Work

### CODE REVIEW

File: bba/bba.go [hunk 1]
```diff
@@ -80,44 +87,117 @@ func (b *BBA) Result() bool {
...
+		aux := &auxRequest{}
+		if err := json.Unmarshal(msg.Bba.Payload, aux); err != nil {
+			return err
+		}
+		return b.handleAuxRequest(id, aux)
 	default:
-		return errors.New(fmt.Sprintf("unexpected request type"))
+		return ErrUndefinedRequestType
+	}
+	return ErrUndefinedRequestType
```
Comment (AgwaB) [inline bba/bba.go]: It is useless return

File: bba/bba.go [hunk 2]
```diff
@@ -80,44 +87,117 @@ func (b *BBA) Result() bool {
...
+	}
+	count := b.countBvalWithValue(bval.Value)
+	if count == 2*b.f+1 {
+		b.binValueSet.union(bval.Value)
+		b.binValueChan <- struct{}{}
+		return nil
+	}
+	if count == b.f+1 && !b.broadcastedBvalSet.exist(bval.Value) {
+		b.broadcastedBvalSet.union(bval.Value)
+		return b.broadcast(pb.BBA_BVAL, bval)
```
Comment (AgwaB) [inline bba/bba.go]: Logic : broadcast `BVAL(est)` and wait until when receive `f+1` message ( if receive `f+1` messages and didn't broadcast, then broadcast)

This looks some logic is skipped!

File: bba/bba_internal_test.go [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+		}
+
+		teardown()
+	}()
+
+	for i, bval := range bvalList {
+		if err := ba.handleBvalRequest(cleisthenes.ConnId(strconv.Itoa(i)), bval); err != nil {
+			t.Fatalf("handle bval request failed with error: %s", err.Error())
+		}
+		if i == 5 {
```
Comment (AgwaB) [inline bba/bba_internal_test.go]: why this is `5`?

File: bba/bba_internal_test.go [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+			if !ba.broadcastedBvalSet.exist(one) {
```
Comment (AgwaB) [inline bba/bba_internal_test.go]: This intended with above `bvalList` init (140 line) ?

File: bba/bba.go [hunk 5]
```diff
@@ -80,44 +87,117 @@ func (b *BBA) Result() bool {
...
+	if err != nil {
+		return err
+	}
+	if r != nil {
+		return nil
+	}
+	return b.bvalRepo.Save(id, data)
+}
+
+func (b *BBA) countBvalWithValue(val Binary) int {
```
Comment (AgwaB) [inline bba/bba.go]: Personally I confuse this name. how about `countBvalByValue`?

File: bba/bba.go [hunk 6]
```diff
@@ -80,44 +87,117 @@ func (b *BBA) Result() bool {
...
 }
 
-func (b *BBA) handleBvalRequest(msg request) error {
+func (b *BBA) handleAuxRequest(id cleisthenes.ConnId, req *auxRequest) error {
 	return nil
 }
 
-func (b *BBA) handleAuxRequest(msg request) error {
+func (b *BBA) broadcast(typ pb.BBAType, req cleisthenes.Request) error {
+	pl, err := json.Marshal(req)
```
Comment (AgwaB) [inline bba/bba.go]: `pl` is too short. Until i see Payload, i didn't understand. I think just `payload` is enough.

File: bba/bba.go (hunk 5 above)
Comment (zeroFruit) [reply bba/bba.go]: good!

File: bba/bba.go [hunk 7]
```diff
@@ -80,44 +87,117 @@ func (b *BBA) Result() bool {
...
+	}
+	return ErrUndefinedRequestType
+}
+
+func (b *BBA) handleBvalRequest(id cleisthenes.ConnId, bval *bvalRequest) error {
+	if err := b.saveBvalIfNotExist(id, bval); err != nil {
+		return err
+	}
+	count := b.countBvalWithValue(bval.Value)
+	if count == 2*b.f+1 {
```
Comment (AgwaB) [inline bba/bba.go]: `2*b.f+1` it is more meaningful if make as variable or function!

File: bba/bba_internal_test.go [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+		result := make([]cleisthenes.Request, 0)
+		for _, req := range bvalRepo.ReqMap {
+			result = append(result, req)
+		}
+		return result
+	}
+	//
+	// broadcaster setup
+	//
+	bc := &mock.Broadcaster{
```
Comment (AgwaB) [inline bba/bba_internal_test.go]: `bc` -> `broadCaster`

**Review by AgwaB — DISMISSED** [review]

**Comment by AgwaB** [comment]
LGTM

**Review by AgwaB — APPROVED** [review]

**Pull request merged**